daml-assistant/src/DAML/Assistant/Cache.hs

```diff
@@ -36,63 +37,112 @@ instance Y.FromJSON UpdateCheck where
+    -> IO [SdkVersion]
+cacheAvailableSdkVersions damlPath getVersions = do
     damlConfigE <- tryConfig $ readDamlConfig damlPath
     let updateCheckM = join $ eitherToMaybe (queryDamlConfig ["update-check"] =<< damlConfigE)
         defaultUpdateCheck = UpdateCheckEvery (CacheTimeout 86400)
     case fromMaybe defaultUpdateCheck updateCheckM of
-        UpdateCheckNever -> pure Nothing
+        UpdateCheckNever -> do
+            valueAgeM <- loadFromCacheWith damlPath versionsKey (CacheTimeout 0) deserializeVersions
+            pure (maybe [] fst valueAgeM)
+
         UpdateCheckEvery timeout ->
-            cacheWith "latest-sdk-version" timeout
-                serializeMaybeSdkVersion deserializeMaybeSdkVersion
-                damlPath getVersion
+            cacheWith damlPath versionsKey timeout
+                serializeVersions deserializeVersions
+                getVersions
 
-serializeMaybeSdkVersion :: Maybe SdkVersion -> String
-serializeMaybeSdkVersion = \case
-    Nothing -> ""
-    Just v -> versionToString v
+serializeVersions :: Serialize [SdkVersion]
+serializeVersions =
+    unlines . map versionToString
 
-deserializeMaybeSdkVersion :: String -> Maybe (Maybe SdkVersion)
-deserializeMaybeSdkVersion = \case
-    "" -> Nothing
-    v  -> fmap Just . eitherToMaybe $ parseVersion (pack v)
+deserializeVersions :: Deserialize [SdkVersion]
+deserializeVersions =
+    Just . mapMaybe (eitherToMaybe . parseVersion . pack) . lines
 
 cacheDirPath :: DamlPath -> FilePath
 cacheDirPath (DamlPath damlPath) = damlPath </> "cache"
 
 cacheFilePath :: DamlPath -> CacheKey -> FilePath
 cacheFilePath damlPath (CacheKey key) = cacheDirPath damlPath </> key
 
+type Serialize t = t -> String
+type Deserialize t = String -> Maybe t
+
 cacheWith
-    :: CacheKey
+    :: DamlPath
+    -> CacheKey
     -> CacheTimeout
-    -> (t -> String)
-    -> (String -> Maybe t)
-    -> DamlPath
+    -> Serialize t
+    -> Deserialize t
     -> IO t
     -> IO t
-cacheWith key (CacheTimeout timeout) serialize deserialize damlPath getValue = do
-    let path = cacheFilePath damlPath key
+cacheWith damlPath key timeout ser deser getFresh = do
+    valueAgeM <- loadFromCacheWith damlPath key timeout deser
+    case valueAgeM of
+        Just (value, Fresh) -> pure value
+        Just (value, Stale) -> do
+            valueE <- tryAny getFresh
+            case valueE of
+                Left _ -> pure value
+                Right value' -> do
+                    saveToCacheWith damlPath key ser value'
+                    pure value'
+        Nothing -> do
+            value <- getFresh
+            saveToCacheWith damlPath key ser value
+            pure value
 
+-- | A representation of the age of a cache value. We only care if the value is stale or fresh.
+data CacheAge
+    = Stale
+    | Fresh
+
+-- | Save value to cache.
+saveToCache :: DamlPath -> CacheKey -> String -> IO ()
+saveToCache damlPath key value =
+    void . tryIO $ do
```

Comment:
Why is catching and discarding all IOExceptions a good idea here? Would be nice to capture this in a comment.

Reply:
I'll add a comment